src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,573 @@
+class sgns_trainer
+{
+  public:
+
+    // Constructor.
+    sgns_trainer(const cpptoml::table& cfg,
+                 const cpptoml::table& embed_cfg,
+                 const cpptoml::table& sgns_cfg) :
+        cfg_(cfg),
+        embed_cfg_(embed_cfg),
+        prefix_(*embed_cfg.get_as<std::string>("prefix")),
+        vector_size_(embed_cfg.get_as<std::size_t>("vector-size").value_or(100)),
+        num_threads_(embed_cfg.get_as<std::size_t>("num-threads").value_or(std::max(1u, std::thread::hardware_concurrency()))),
+        max_ram_(embed_cfg.get_as<std::size_t>("max-ram").value_or(4096) * 1024 * 1024),
+        subsample_threshold_(sgns_cfg.get_as<double>("subsample-threshold").value_or(1E-4)),
+        max_window_size_(sgns_cfg.get_as<std::size_t>("max-window-size").value_or(6)),
+        iterations_(sgns_cfg.get_as<uint64_t>("iterations").value_or(10)),
+        starting_learning_rate_((float)sgns_cfg.get_as<double>("learning-rate").value_or(0.025)),
+        negative_samples_(sgns_cfg.get_as<std::size_t>("negative-samples").value_or(20)),
+        binary_output_(sgns_cfg.get_as<bool>("binary").value_or(true)),
+        vocab_(load_vocab(prefix_ + "/vocab.bin")),
+        noise_dist_(create_unigram_noise_distribution(1E8)), // TODO: Configurable size?
+        net_(vocab_.vector.size(), vector_size_),
+        learning_rate_(starting_learning_rate_),
+        word_count_actual_(0)
+    {
+        std::mutex io_mutex;
+        printing::progress progress{" > Training: ", (uint64_t)(starting_learning_rate_ * 100000.0)};
+
+        train(io_mutex, progress);
+
+        save_w2v_vectors();
+        save_meta_vectors();
+    }
+
+    // When the thread_pool goes out of scope, the threads will join. In other
+    // words, when this method returns, all threads have finished training.
+    void train(std::mutex& io_mutex, printing::progress& progress)
+    {
+        auto stream = analyzers::load_filters(cfg_, embed_cfg_);
+        if (!stream)
+        {
+            throw sgns_exception{"Failed to find an ngram-word analyzer configuration"};
+        }
+
+        parallel::thread_pool pool{num_threads_};
+
+        auto iterations_left = iterations_;
+
+        auto docs = corpus::make_corpus(cfg_);
+
+        while (iterations_left > 0)
+        {
+            std::uniform_int_distribution<std::size_t> next_random;
+            int64_t word_counter = 0;
+
+            corpus::parallel_consume(
+                *docs, pool,
+                [&]() {
+                    return sgns_local_buffer{*stream};
+                },
+                [&](sgns_local_buffer& buffer, const corpus::document& doc) {
+                    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+                    sgns_net_vector neu1e(vector_size_, 0);        
+                    std::default_random_engine engine(seed);
+                    std::uniform_int_distribution<std::size_t> next_random;
```

Comment:
Actually, it might be good to move these into your `sgns_local_buffer` to avoid extra allocations/work occurring every time a new document is being processed.